[Pfc] ResourceMonitor and PurgePlugin

osschar opened the pull request:
Preliminary PR -- a place holder to get this into 5.8.

Some minimal changes still expected:
- improved management of usage snap-shots;
- optimization of age-based purge decision to use stat instead of unpacking cinfo to determine the last access time;
- potential issues while testing in production environment.

amadio committed:
Second rebase of alja/purge-main-rb1 onto master / 5.7.2.

osschar committed:
Implement purging of empty leaf directories. Prepare for cleaner snapshot/purgeshot generation.

osschar committed:
Ensure snap-shots get correct equidistant stats and purge-shots get freshly updated usages.

alja committed:
Use cinfo modification time instead of opening and reding the cinfo file.

alja committed:
Correction in  DataFsState::reset_sshot_stats

alja committed:
Uncomment DirStat dump to json in ResourceMonitor heart beat loop. Minor corrections in DataFsSnapshot::write_json_file.

alja commented on src/XrdPfc.cmake:
```diff
@@ -15,11 +16,18 @@ set( LIB_XRD_BLACKLIST  XrdBlacklistDecision-${PLUGIN_VERSION} )
 add_library(
   ${LIB_XRD_FILECACHE}
-  MODULE
+  SHARED
```
Thanks Brian! I will look into this.

osschar replied:
Yeah, this is all my fault again -- I admit I didn't really know what I'm doing, just trying to get it to work :) I did this so we would be able to link the purge-plugin against the pfc ... it's mostly about the data-formats and the (virtual, probably) base-class. I think we can work around by just isolating all this in a header-only files.

I was sort of surprised that the linker complained about this -- it used to be that shared libs didn't require all the symbols to be defined at link time -- assuming they will be satisfied at dynamic link time, ie, when loading the library into a running executable.

bbockelm replied:
Depends on the linker!  I believe what you write is basically correct on Linux (though games can be played with symbol hiding -- you're right for most out-of-the-box items).

CMake will refuse to let you to define a build where you link against a module.  I think Windows, perhaps (or some other platform I'm not familiar with), makes a real hard distinction between modules versus shared libraries.

I only noticed this on Mac OS X because it causes the suffix to change from `.so` to `.dylib` and XRootD only searches for `.so` files (meaning everything broke!).

As a workaround, I patched the following in:
```
--- a/src/XrdPfc.cmake
+++ b/src/XrdPfc.cmake
@@ -38,6 +38,8 @@ add_library(
   XrdPfc/XrdPfcIOFileBlock.cc   XrdPfc/XrdPfcIOFileBlock.hh
   XrdPfc/XrdPfcDecision.hh)
 
+set_target_properties( ${LIB_XRD_FILECACHE} PROPERTIES SUFFIX ".so" )
+
 target_link_libraries(
   ${LIB_XRD_FILECACHE}
   PRIVATE
```

but this is really the wrong solution.  @osschar - I think your suggestion of header-only classes/structs is the best approach.

osschar committed:
Rework res-mon updates of st_blocks used to work with aggressive block preallocation.

- Never report st_blocks usage beyond the known target file-size rounded up to nearest 4 kB.

- Break-up File destruction so one can call a final stat after the file has
  been closed -- at this point the st_blocks is correct. Report any final
  discrepancies before emitting the close record to res-mon.

- This required some gimanstics in Cace::dec_ref_cnt() and
  Cache::UnlinkFile(force). Additionally, a nullptr is put into the active map
  to prevent the file from being re-opened before this final stat is complete.

- Some other minor tweaks / cleanups.

osschar committed:
In forced-evict make sure we do not over report purged st_blocks now that
we limit reporting in stat-updates.

osschar committed:
Make all purge-plugin dataformats header-only, revert libXrdPfc lib type to MODULE, cleanup some includes.

osschar commented:
Anyone has a clue why the CI builds fail with a supposedly missing include file? Is it trying to do an incremental build?
```
In file included from /usr/include/xrootd/XrdPfc/XrdPfcDirStateSnapshot.hh:4:
28 /usr/include/xrootd/XrdPfc/XrdPfcDirState.hh:5:10: fatal error: XrdPfcDirStateBase.hh: No such file or directory
29     5 | #include "XrdPfcDirStateBase.hh"
30       |          ^~~~~~~~~~~~~~~~~~~~~~~
31 compilation terminated.
32 /usr/include/xrootd/XrdPfc/XrdPfcDirState.hh:5:10: fatal error: XrdPfcDirStateBase.hh: No such file or directory
33     5 | #include "XrdPfcDirStateBase.hh"
34       |          ^~~~~~~~~~~~~~~~~~~~~~~
35 compilation terminated.
```

osschar commented:
Oh, it's in post-build --- I guess I need to add some of the new headers into the exported section.

osschar committed:
Export the new headers relevant for purge-plugin, remove the previous ones.

osschar committed:
Export also the XrdPfcPathParseTools.hh.

amadio commented:
> Oh, it's in post-build --- I guess I need to add some of the new headers into the exported section.

Likely there is a public header including a private one. You can check locally with the `test/check-headers.sh` script. It just checks that each installed header can be included without errors.

alja committed:
Use stat's m_time macro to access st_mtim.tv_sec.

alja committed:
Correction in calculating space to remove when disk usage exceed high limit and is below low watermark

alja committed:
Add dirstat report interval in configuration.

osschar committed:
Minor cleanup of snapshot reports.

bbockelm commented:
@alja - there are merge conflicts that have evolved - can you fix?

osschar commented:
Oh, crap ... i made a new branch on Thursday ... but forgot i have to force-push it here :(
I'll make a new PR with a more generic branch name and close this one :)

osschar commented:
Superseded by #2436.
